components/ntask/org.wso2.carbon.ntask.core/src/main/java/org/wso2/carbon/ntask/core/impl/AbstractQuartzTaskManager.java

```diff
@@ -183,29 +202,50 @@ protected synchronized void scheduleAllTasks() throws TaskException {
-        boolean paused = TaskUtils.isTaskPaused(this.getTaskRepository(), taskName);
-        this.scheduleLocalTask(taskName, paused);
+        scheduleLocalTask(taskName, false);
     }
 
+    /**
+     * Method to schedule a task with whether or not the task is paused and/or requires recovery set to the specified
+     * values.
+     * @param taskName the name of the task to be scheduled
+     * @param paused whether or not the task is paused
+     * @param requestRecovery whether or not the task requires recovery
+     * @throws TaskException if there is an error in scheduling the task
+     */
     protected synchronized void scheduleLocalTask(String taskName,
-                                                  boolean paused) throws TaskException {
+                                                  boolean paused,
+                                                  boolean requestRecovery) throws TaskException {
         TaskInfo taskInfo = this.getTaskRepository().getTask(taskName);
         String taskGroup = this.getTenantTaskGroup();
         if (taskInfo == null) {
             throw new TaskException("Non-existing task for scheduling with name: " + taskName,
                     Code.NO_TASK_EXISTS);
         }
-        if (this.containsLocalTask(taskName, taskGroup)) {
-            /* to make the scheduleLocalTask operation idempotent */
-            return;
-        }
         Class<? extends Job> jobClass = taskInfo.getTriggerInfo().isDisallowConcurrentExecution() ?
                 NonConcurrentTaskQuartzJobAdapter.class : TaskQuartzJobAdapter.class;
-        JobDetail job = JobBuilder.newJob(jobClass).withIdentity(taskName, taskGroup).usingJobData(
-                this.getJobDataMapFromTaskInfo(taskInfo)).build();
+        if (!taskInfo.getProperties().containsKey(TaskConstants.REQUEST_RECOVERY)) {
+            Map<String, String> taskPropertiesMap = taskInfo.getProperties();
+            taskPropertiesMap.put(TaskConstants.REQUEST_RECOVERY, Boolean.toString(requestRecovery));
+            taskInfo.setProperties(taskPropertiesMap);
+        }
+        JobDetail job = JobBuilder.newJob(jobClass).withIdentity(taskName, taskGroup).requestRecovery(requestRecovery).
+                usingJobData(this.getJobDataMapFromTaskInfo(taskInfo)).build();
         Trigger trigger = this.getTriggerFromInfo(taskName, taskGroup, taskInfo.getTriggerInfo());
         try {
+            //Deletion is done to replace the task if already scheduled.
+            if (this.getScheduler().checkExists(new JobKey(taskName, taskGroup))) {
+                this.getScheduler().deleteJob(new JobKey(taskName, taskGroup));
+                if (log.isDebugEnabled()) {
+                    log.debug("Deleted task [" + taskName + "] to override.");
+                }
+            }
             this.getScheduler().scheduleJob(job, trigger);
             if (paused) {
                 this.getScheduler().pauseJob(job.getKey());
```

Comment:
If the task is triggered before pausing, ESB will run a paused task.

Reply:
Yes, it seems with the previous implementation we had to follow this approach. With the new implementation, we may be able to avoid this, since the task and task state are maintained in the shared database. Looking into this.